OpenSim/Common/Component.cpp

```diff
@@ -838,26 +836,61 @@ void Component::constructOutputForStateVariable(const std::string& name)
             SimTK::Stage::Model);
 }
 
+// mark components owned as properties as subcomponents
+void Component::markPropertiesAsSubcomponents()
+{
+    // if being invoked either constructing a new Component
+    // or the properties have been modified. In the latter case
+    // we must make sure that pointers to old properties are cleared
+    _components.clear();
+
+    // Now mark properties that are Components as subcomponents
+    for (int i = 0; i < getNumProperties(); ++i) {
+        auto& prop = updPropertyByIndex(i);
+        if (prop.isObjectProperty()) {
+            for (int j = 0; j < prop.size(); ++j) {
+                Object& obj = prop.updValueAsObject(j);
+                if (Component* comp = dynamic_cast<Component*>(&obj) ) {
+                    markAsSubcomponent(comp);
+                }
+                //Set<Component>* objects = dynamic_cast<Set<Component>*>(&obj)
+                else {
+                    std::string objType = obj.getConcreteClassName();
+                    if (obj.hasProperty("objects")) {
+                        auto& objectsProp = obj.updPropertyByName("objects");
+                        for (int k = 0; k < objectsProp.size(); ++k) {
+                            Object& obj = objectsProp.updValueAsObject(k);
+                            if (Component* comp = dynamic_cast<Component*>(&obj) )
+                                markAsSubcomponent(comp);
```

Comment:
The reason I go one level down is because the Set property has another level of "objects", which are the actual subcomponents. The subcomponents of those components are added when those components are asked to finalizeFromProperties() which is propogated down its subtree of components (of course the marking as subcomponents is the first thing to happen in finalizeFromProperties). Does that make more sense?